test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriBuilder/ContextUrlPathStringTests.cs

```diff
@@ -17,51 +18,253 @@ public class ContextUrlPathStringTests : UriBuilderTestBase
-            Assert.Equal("City", res);
+            string res = this.GetContextUrlPathString(HardCodedTestModel.TestModel, queryUri);
+            Assert.Equal("Edm.String", res);
         }
 
         [Fact]
         public void ContextUrlPathWithEntityServiceOperationIsComposable()
         {
             Uri queryUri = new Uri("GetCoolestPerson/Fully.Qualified.Namespace.Employee", UriKind.Relative);
-            string res = this.GetContextUrlPathString(queryUri);
-            Assert.Equal("Fully.Qualified.Namespace.Employee", res);
+            string res = this.GetContextUrlPathString(HardCodedTestModel.TestModel, queryUri);
+            Assert.Equal("People/Fully.Qualified.Namespace.Employee", res);
         }
 
-        #region private methods
-        private string GetContextUrlPathString(Uri queryUri)
+        [Theory]
+        [InlineData("People(1)/Fully.Qualified.Namespace.GetSomeAddressFromPerson()/Street")]
+        [InlineData("People(1)/Fully.Qualified.Namespace.GetSomeAddressFromPerson/Street")]
+        public void ContextUrlPathWithComplexPropertyAccessAfterOperationIsComposable(string uri)
         {
-            ODataUriParser odataUriParser = new ODataUriParser(HardCodedTestModel.TestModel, queryUri);
+            Uri queryUri = new Uri(uri, UriKind.Relative);
+            string res = this.GetContextUrlPathString(HardCodedTestModel.TestModel, queryUri);
+            Assert.Equal("Edm.String", res);
+        }
+
+
+        [Theory]
+        [InlineData("Customers(1)/NS.GetSomeOrders")]
+        [InlineData("Customers(1)/NS.GetSomeOrders()")]
+        [InlineData("Customers(1)/NS.GetAnOrder")]
+        [InlineData("Customers(1)/NS.GetAnOrder()")]
+        public void ContextUrlPathWithOperationWithEntitySetPathAndReturnType(string uri)
+        {
+            IEdmModel model = GetEdmModel();
+            Uri queryUri = new Uri(uri, UriKind.Relative);
+            string res = this.GetContextUrlPathString(model, queryUri);
+            Assert.Equal("Customers(1)/Orders", res);
+        }
+
+        [Fact]
+        public void ContextUrlPathWithOperationWithoutEntitySetPathAndComplexReturnType()
+        {
+            IEdmModel model = GetEdmModel();
+            Uri queryUri = new Uri("Customers(1)/NS.GetSomeAddressFromCustomer1", UriKind.Relative);
+            string res = this.GetContextUrlPathString(model, queryUri);
+            Assert.Equal("NS.Address", res);
+        }
+
+        [Fact]
+        public void ContextUrlPathWithPropertyAccessAfterOperationWithoutEntitySetPathAndComplexReturnType()
+        {
+            IEdmModel model = GetEdmModel();
+            Uri queryUri = new Uri("Customers(1)/NS.GetSomeAddressFromCustomer3/City", UriKind.Relative);
+            string res = this.GetContextUrlPathString(model, queryUri);
+            Assert.Equal("Edm.String", res);
+        }
+
+        [Fact]
+        public void ContextUrlPathWithPropertyAccessOperationWithEntitySetPathAndComplexReturnTypeThrows()

```

Comment:
>ContextUrlPathWithPropertyAccessOperationWithEntitySetPathAndComplexReturnTypeThrows [](start = 20, length = 84)

This is really an issue with the model, right?  The binding path is not consistent with the return type?  Shouldn't we (ideally) catch this when we load the model?  I'm not suggesting you do that code now, but I'd rather have the check there than catch the exception here. Perhaps remove this test and file a bug to validate the entityset path when building a model? Or you can just put in the bug that this test should be removed once that test is written?

Reply:
Filed as bug at: https://github.com/OData/odata.net/issues/1029